In [1]:
import pandas as pd

In [2]:
match =pd.read_parquet("data/processed/match.parquet")
odds =pd.read_parquet("data/processed/odds.parquet")
point_by_point =pd.read_parquet("data/processed/point_by_point.parquet")
statistics =pd.read_parquet("data/processed/statistics.parquet")
tennis_power =pd.read_parquet("data/processed/tennis_power.parquet")
votes =pd.read_parquet("data/processed/votes.parquet")

In [3]:
type(match)

pandas.core.frame.DataFrame

In [4]:
datasets={
    "match":match,
    "odds":odds,
    "point_by_point":point_by_point,
    "statistics":statistics,
    "tennis_power":tennis_power,
    "votes":votes
}
for name,df in datasets.items():
    print(name,type(df))

match <class 'pandas.core.frame.DataFrame'>
odds <class 'pandas.core.frame.DataFrame'>
point_by_point <class 'pandas.core.frame.DataFrame'>
statistics <class 'pandas.core.frame.DataFrame'>
tennis_power <class 'pandas.core.frame.DataFrame'>
votes <class 'pandas.core.frame.DataFrame'>


In [5]:
for name,df in datasets.items():
    print("\n",name)
    print(df.isna().sum().sort_values(ascending=False))


 match
normal_time           316802
period_4              316802
period_4_tie_break    316802
period_5_tie_break    316802
period_5              316802
                       ...  
period_2              230817
period_1              230061
name                  212035
match_id                   0
date                       0
Length: 63, dtype: int64

 odds
winnig                      3900
match_id                       0
market_id                      0
market_name                    0
is_live                        0
suspended                      0
initial_fractional_value       0
fractional_value               0
choice_name                    0
choice_source_id               0
change                         0
date                           0
dtype: int64

 point_by_point
match_id             0
set_id               0
game_id              0
point_id             0
home_point           0
away_point           0
point_description    0
home_point_type      0
away_point_type      0
home_sco

## Missing Values

After checking the missing values in all DataFrames, the following observations were found:

- The `match` DataFrame contains a large number of missing values across several columns.
- In the `odds` DataFrame, the `winning` column contains missing values.
- In the `statistics` DataFrame, the `home_total` and `away_total` columns contain missing values.

These missing values need to be investigated before deciding whether to remove them, replace them, or keep them.

In [6]:
for name,df in datasets.items():
    print(name,df.duplicated().sum())

match 5870
odds 0
point_by_point 0
statistics 0
tennis_power 0
votes 0


Duplicate records should be investigated because they may distort the analysis. However, they should not be removed automatically before understanding what one row represents in each DataFrame.


In [7]:
for name,df in datasets.items():
    print(name)
    print("Row:",df.shape[0])
    print("Columns:",df.shape[1])
    print("-----------------------")

match
Row: 316802
Columns: 63
-----------------------
odds
Row: 60946
Columns: 12
-----------------------
point_by_point
Row: 2549369
Columns: 14
-----------------------
statistics
Row: 1358234
Columns: 14
-----------------------
tennis_power
Row: 469677
Columns: 6
-----------------------
votes
Row: 35658
Columns: 4
-----------------------


In [8]:
match.head()


,match_id,name,slug,gender,user_count,residence,birthplace,height,weight,plays,...,has_event_player_statistics,crowd_sourcing_enabled,has_performance_graph_feature,display_inverse_home_away_teams,priority,competition_type,city,stadium,venue_id,date
0,11998445,Auger-Aliassime F.,auger-aliassime-felix,M,23318.0,"Monte Carlo, Monaco","Montreal, Canada",1.93,87.0,right-handed,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
1,11998446,Cobolli F.,flavio-cobolli,M,5995.0,Rome,"Florence, Italy",1.83,71.0,right-handed,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
2,11998447,Martínez P.,martinez-pedro,M,2621.0,"Valencia, Spain","Alzira, Spain",1.85,76.0,right-handed,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
3,11998448,Muller A.,muller-alexandre,M,2443.0,France,"Poissy, France",1.83,73.0,right-handed,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
4,11998449,Mayot H.,mayot-harold,M,1492.0,"Marly, France","Metz, France",1.78,78.0,right-handed,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201


In [9]:
match.columns


Index(['match_id', 'name', 'slug', 'gender', 'user_count', 'residence',
       'birthplace', 'height', 'weight', 'plays', 'turned_pro',
       'current_prize', 'total_prize', 'player_id', 'current_rank',
       'name_code', 'country', 'full_name', 'current_score', 'display_score',
       'period_1', 'period_2', 'period_3', 'period_4', 'period_5',
       'period_1_tie_break', 'period_2_tie_break', 'period_3_tie_break',
       'period_4_tie_break', 'period_5_tie_break', 'normal_time',
       'first_to_serve', 'home_team_seed', 'away_team_seed', 'custom_id',
       'winner_code', 'default_period_count', 'start_datetime', 'match_slug',
       'final_result_only', 'round_id', 'cup_round_type', 'season_id', 'year',
       'current_period_start_timestamp', 'tournament_id', 'tournament_name',
       'tournament_slug', 'tournament_unique_id', 'tournament_category_name',
       'tournament_category_slug', 'ground_type', 'tennis_points',
       'has_event_player_statistics', 'crowd_sourcing_enabl

In [21]:
odds.columns

Index(['match_id', 'market_id', 'market_name', 'is_live', 'suspended',
       'initial_fractional_value', 'fractional_value', 'choice_name',
       'choice_source_id', 'winnig', 'change', 'date'],
      dtype='object')

In [22]:
statistics.columns

Index(['match_id', 'period', 'statistic_category_name', 'statistic_name',
       'home_stat', 'away_stat', 'compare_code', 'statistic_type',
       'value_type', 'home_value', 'away_value', 'home_total', 'away_total',
       'date'],
      dtype='object')

In [23]:
tennis_power.columns

Index(['match_id', 'set_num', 'game_num', 'value', 'break_occurred', 'date'], dtype='object')

In [24]:
point_by_point.columns

Index(['match_id', 'set_id', 'game_id', 'point_id', 'home_point', 'away_point',
       'point_description', 'home_point_type', 'away_point_type', 'home_score',
       'away_score', 'serving', 'scoring', 'date'],
      dtype='object')

In [25]:
votes.columns

Index(['match_id', 'home_vote', 'away_vote', 'date'], dtype='object')

In [10]:
match["match_id"].nunique()

16873

In [11]:
len(match)

316802

In [12]:
duplictes = match[match.duplicated(keep=False)]  #keep=False is useful because it shows both the original row and its copies
duplictes.shape

(8882, 63)

In [13]:
duplictes.sort_values("match_id").head(30)

,match_id,name,slug,gender,user_count,residence,birthplace,height,weight,plays,...,has_event_player_statistics,crowd_sourcing_enabled,has_performance_graph_feature,display_inverse_home_away_teams,priority,competition_type,city,stadium,venue_id,date
337,12011297,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
1180,12011297,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
2012,12011297,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
339,12011306,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
1182,12011306,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240201
35969,12018506,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240209
35290,12018506,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240209
32133,12018506,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240208
33102,12018506,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240208
35347,12018734,None,None,None,NaN,None,None,NaN,NaN,None,...,None,None,None,None,NaN,NaN,None,None,NaN,20240209


In [14]:
print("Total rows:", len(match))
print("Unique match_id:", match["match_id"].nunique())
print("Exact duplicate rows:", match.duplicated().sum())
print("Duplicated match_id:", match["match_id"].duplicated().sum())

Total rows: 316802
Unique match_id: 16873
Exact duplicate rows: 5870
Duplicated match_id: 299929


### Investigation of Duplicate Match IDs

The `match` DataFrame contains 316,802 rows but only 16,873 unique `match_id` values. This indicates that multiple rows can belong to the same tennis match.

Therefore, repeated `match_id` values should not automatically be considered duplicate records.

There are also 5,870 exact duplicate rows, where all column values are identical. These records require further investigation before removal.

The next step is to examine multiple rows belonging to the same `match_id` to understand what each row represents.

In [15]:
summary=[]
for col in match.columns:
    summary.append([col,match[col].dtypes,match[col].nunique(),match[col].isna().sum(),round(match[col].isna().mean()*100.2)])
summary =pd.DataFrame(summary,columns=["columns","dtype","unique_values","missing","missing_percent"])
summary

,columns,dtype,unique_values,missing,missing_percent
0,match_id,int64,16873,0,0
1,name,object,2933,212035,67
2,slug,object,2658,247706,78
3,gender,object,2,267057,84
4,user_count,float64,4511,231318,73
...,...,...,...,...,...
58,competition_type,float64,2,281616,89
59,city,object,190,281379,89
60,stadium,object,250,281379,89
61,venue_id,float64,939,281379,89


In [16]:
statistics.head()

,match_id,period,statistic_category_name,statistic_name,home_stat,away_stat,compare_code,statistic_type,value_type,home_value,away_value,home_total,away_total,date
0,11998445,ALL,service,aces,12,6,1,positive,event,12,6,NaN,NaN,20240201
1,11998445,ALL,service,double_faults,2,7,2,negative,event,2,7,NaN,NaN,20240201
2,11998445,ALL,service,first_serve,57/101 (56%),53/90 (59%),2,positive,team,57,53,101.0,90.0,20240201
3,11998445,ALL,service,second_serve,42/44 (95%),30/37 (81%),1,positive,team,42,30,44.0,37.0,20240201
4,11998445,ALL,service,first_serve_points,42/57 (74%),39/53 (74%),1,positive,team,42,39,57.0,53.0,20240201


In [17]:
odds.head()

,match_id,market_id,market_name,is_live,suspended,initial_fractional_value,fractional_value,choice_name,choice_source_id,winnig,change,date
0,11974053,1,full_time,False,False,1/4,1/5,1,1994690951,True,-1,20240201
1,11974053,1,full_time,False,False,11/4,10/3,2,1994691017,False,1,20240201
2,11974066,1,full_time,False,False,14/1,14/1,1,1990515041,False,0,20240201
3,11974066,1,full_time,False,False,1/33,3/100,2,1990515081,True,0,20240201
4,11998445,1,full_time,False,False,1/1,4/5,1,1985921173,False,-1,20240201


In [18]:
point_by_point.head()

,match_id,set_id,game_id,point_id,home_point,away_point,point_description,home_point_type,away_point_type,home_score,away_score,serving,scoring,date
0,11998445,3,13,0,1,0,0,6,5,6,7,1,2,20240201
1,11998445,3,13,1,1,1,0,5,6,6,7,1,2,20240201
2,11998445,3,13,2,1,2,0,5,6,6,7,1,2,20240201
3,11998445,3,13,3,1,3,0,5,1,6,7,1,2,20240201
4,11998445,3,13,4,1,4,0,5,1,6,7,1,2,20240201


In [19]:
votes.head()

,match_id,home_vote,away_vote,date
0,11974053,272,61,20240201
1,11974066,124,579,20240201
2,11998445,4740,4091,20240201
3,11998446,897,4558,20240201
4,11998447,4456,780,20240201


In [20]:
tennis_power.head()

,match_id,set_num,game_num,value,break_occurred,date
0,11998445,1,1,-52.80,False,20240201
1,11998445,1,2,48.14,False,20240201
2,11998445,1,3,-51.62,False,20240201
3,11998445,1,4,10.00,False,20240201
4,11998445,1,5,26.60,True,20240201
